File: golang/direct_e2ee/ratchet.go
package directe2ee

import (
	"crypto/ecdh"

	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/hkdf"
)

const MaxSkip uint32 = 1000

// ChainStep contains derived message material for a single ratchet step.
type ChainStep struct {
	MessageKey   [32]byte
	Nonce        [12]byte
	NextChainKey [32]byte
}

// RootStep contains root and chain material derived from a DH ratchet step.
type RootStep struct {
	RootKey  [32]byte
	ChainKey [32]byte
}

// DeriveChainStep implements P5 kdf_ck(CK) -> (CK', MK, NONCE).
func DeriveChainStep(chainKey [32]byte) ChainStep {
	prk := hkdf.Extract(hashProvider, chainKey[:], make([]byte, 32))
	out, err := hkdfExpandOnly(prk, []byte("ANP Direct E2EE v1 KDF_CK"), 76)
	if err != nil {
		panic(err)
	}
	var nextChainKey [32]byte
	var messageKey [32]byte
	var nonce [12]byte
	copy(nextChainKey[:], out[0:32])
	copy(messageKey[:], out[32:64])
	copy(nonce[:], out[64:76])
	return ChainStep{MessageKey: messageKey, Nonce: nonce, NextChainKey: nextChainKey}
}

// DeriveRootStep implements P5 kdf_rk(RK, dh_out) -> (RK', CK).
func DeriveRootStep(rootKey [32]byte, dhOut []byte) (RootStep, error) {
	prk := hkdf.Extract(hashProvider, dhOut, rootKey[:])
	out, err := hkdfExpandOnly(prk, []byte("ANP Direct E2EE v1 KDF_RK"), 64)
	if err != nil {
		return RootStep{}, err
	}
	var nextRoot [32]byte
	var chain [32]byte
	copy(nextRoot[:], out[0:32])
	copy(chain[:], out[32:64])
	return RootStep{RootKey: nextRoot, ChainKey: chain}, nil
}

func deriveRootStepFromKeys(rootKey [32]byte, privateKey *ecdh.PrivateKey, publicKey *ecdh.PublicKey) (RootStep, error) {
	dhOut, err := privateKey.ECDH(publicKey)
	if err != nil {
		return RootStep{}, err
	}
	return DeriveRootStep(rootKey, dhOut)
}

// EncryptWithStep encrypts plaintext with the derived step.
func EncryptWithStep(step ChainStep, plaintext []byte, aad []byte) ([]byte, error) {
	aead, err := chacha20poly1305.New(step.MessageKey[:])
	if err != nil {
		return nil, cryptoError("invalid ChaCha20-Poly1305 key")
	}
	return aead.Seal(nil, step.Nonce[:], plaintext, aad), nil
}

// DecryptWithStep decrypts ciphertext with the derived step.
func DecryptWithStep(step ChainStep, ciphertext []byte, aad []byte) ([]byte, error) {
	aead, err := chacha20poly1305.New(step.MessageKey[:])
	if err != nil {
		return nil, cryptoError("invalid ChaCha20-Poly1305 key")
	}
	plaintext, err := aead.Open(nil, step.Nonce[:], ciphertext, aad)
	if err != nil {
		return nil, cryptoError("failed to decrypt ciphertext")
	}
	return plaintext, nil
}

File: rust/src/direct_e2ee/ratchet.rs
use super::errors::DirectE2eeError;
use super::x3dh::{hkdf_expand_prk, hkdf_extract};
use ring::aead::{Aad, LessSafeKey, Nonce, UnboundKey, CHACHA20_POLY1305};

pub const MAX_SKIP: u32 = 1000;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ChainStep {
    pub message_key: [u8; 32],
    pub nonce: [u8; 12],
    pub next_chain_key: [u8; 32],
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RootStep {
    pub root_key: [u8; 32],
    pub chain_key: [u8; 32],
}

pub fn derive_chain_step(chain_key: &[u8; 32]) -> ChainStep {
    let prk = hkdf_extract(&[0u8; 32], chain_key);
    let out = hkdf_expand_prk(&prk, b"ANP Direct E2EE v1 KDF_CK", 76)
        .expect("hkdf expand should support fixed length");
    let mut next_chain_key = [0u8; 32];
    let mut message_key = [0u8; 32];
    let mut nonce = [0u8; 12];
    next_chain_key.copy_from_slice(&out[0..32]);
    message_key.copy_from_slice(&out[32..64]);
    nonce.copy_from_slice(&out[64..76]);
    ChainStep {
        message_key,
        nonce,
        next_chain_key,
    }
}

pub fn derive_root_step(root_key: &[u8; 32], dh_out: &[u8]) -> Result<RootStep, DirectE2eeError> {
    let prk = hkdf_extract(root_key, dh_out);
    let out = hkdf_expand_prk(&prk, b"ANP Direct E2EE v1 KDF_RK", 64)?;
    let mut next_root = [0u8; 32];
    let mut chain_key = [0u8; 32];
    next_root.copy_from_slice(&out[0..32]);
    chain_key.copy_from_slice(&out[32..64]);
    Ok(RootStep {
        root_key: next_root,
        chain_key,
    })
}

pub fn encrypt_with_step(
    step: &ChainStep,
    plaintext: &[u8],
    aad: &[u8],
) -> Result<Vec<u8>, DirectE2eeError> {
    let unbound = UnboundKey::new(&CHACHA20_POLY1305, &step.message_key)
        .map_err(|_| DirectE2eeError::crypto("invalid ChaCha20-Poly1305 key"))?;
    let key = LessSafeKey::new(unbound);
    let nonce = Nonce::assume_unique_for_key(step.nonce);
    let mut buffer = plaintext.to_vec();
    key.seal_in_place_append_tag(nonce, Aad::from(aad), &mut buffer)
        .map_err(|_| DirectE2eeError::crypto("failed to encrypt ciphertext"))?;
    Ok(buffer)
}

pub fn decrypt_with_step(
    step: &ChainStep,
    ciphertext: &[u8],
    aad: &[u8],
) -> Result<Vec<u8>, DirectE2eeError> {
    let unbound = UnboundKey::new(&CHACHA20_POLY1305, &step.message_key)
        .map_err(|_| DirectE2eeError::crypto("invalid ChaCha20-Poly1305 key"))?;
    let key = LessSafeKey::new(unbound);
    let nonce = Nonce::assume_unique_for_key(step.nonce);
    let mut buffer = ciphertext.to_vec();
    let plaintext = key
        .open_in_place(nonce, Aad::from(aad), &mut buffer)
        .map_err(|_| DirectE2eeError::crypto("failed to decrypt ciphertext"))?;
    Ok(plaintext.to_vec())
}

#[cfg(test)]
mod tests {
    use super::{decrypt_with_step, derive_chain_step, encrypt_with_step};
    #[test]
    fn chain_step_encrypts_and_decrypts() {
        let step = derive_chain_step(&[9u8; 32]);
        let ciphertext = encrypt_with_step(&step, b"hello", br#"{"aad":true}"#).expect("encrypt");
        let plaintext = decrypt_with_step(&step, &ciphertext, br#"{"aad":true}"#).expect("decrypt");
        assert_eq!(plaintext, b"hello");
    }
}
